src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1010 @@
+
+  -- | @uniformWord64R upperBound g@ generates a 'Word64' that is uniformly
+  -- distributed over the range @[0, upperBound]@.
+  --
+  -- @since 1.2
+  uniformWord64R :: Word64 -> g -> m Word64
+  uniformWord64R = unsignedBitmaskWithRejectionM uniformWord64
+
+  -- | Generates a 'Word8' that is uniformly distributed over the entire 'Word8'
+  -- range.
+  --
+  -- The default implementation extracts a 'Word8' from 'uniformWord32'.
+  --
+  -- @since 1.2
+  uniformWord8 :: g -> m Word8
+  uniformWord8 = fmap fromIntegral . uniformWord32
+
+  -- | Generates a 'Word16' that is uniformly distributed over the entire
+  -- 'Word16' range.
+  --
+  -- The default implementation extracts a 'Word16' from 'uniformWord32'.
+  --
+  -- @since 1.2
+  uniformWord16 :: g -> m Word16
+  uniformWord16 = fmap fromIntegral . uniformWord32
+
+  -- | Generates a 'Word32' that is uniformly distributed over the entire
+  -- 'Word32' range.
+  --
+  -- The default implementation extracts a 'Word32' from 'uniformWord64'.
+  --
+  -- @since 1.2
+  uniformWord32 :: g -> m Word32
+  uniformWord32 = fmap fromIntegral . uniformWord64
+
+  -- | Generates a 'Word64' that is uniformly distributed over the entire
+  -- 'Word64' range.
+  --
+  -- The default implementation combines two 'Word32' from 'uniformWord32' into
+  -- one 'Word64'.
+  --
+  -- @since 1.2
+  uniformWord64 :: g -> m Word64
+  uniformWord64 g = do
+    l32 <- uniformWord32 g
+    h32 <- uniformWord32 g
+    pure (unsafeShiftL (fromIntegral h32) 32 .|. fromIntegral l32)
+
+  -- | @uniformShortByteString n g@ generates a 'ShortByteString' of length @n@
+  -- filled with pseudo-random bytes.
+  --
+  -- @since 1.2
+  uniformShortByteString :: Int -> g -> m ShortByteString
+  default uniformShortByteString :: MonadIO m => Int -> g -> m ShortByteString
+  uniformShortByteString n = genShortByteStringIO n . uniformWord64
+  {-# INLINE uniformShortByteString #-}
+
+
+
+-- | This class is designed for stateful RNGs that can be saved as immutable data type.
+--
+-- @since 1.2
+class StatefulGen (MutableGen f m) m => FrozenGen f m where
+  -- | Represents the state of the pseudo-random number generator for use with
+  -- 'thawGen' and 'freezeGen'.
+  --
+  -- @since 1.2
+  type MutableGen f m = (g :: Type) | g -> f
+  -- | Saves the state of the pseudo-random number generator as a frozen 'Seed'.
+  --
+  -- @since 1.2
+  freezeGen :: MutableGen f m -> m f
+  -- | Restores the pseudo-random number generator from its frozen 'Seed'.
+  --
+  -- @since 1.2
+  thawGen :: f -> m (MutableGen f m)
+
+
+data MBA s = MBA (MutableByteArray# s)
+
+
+-- | Efficiently generates a sequence of pseudo-random bytes in a platform
+-- independent manner. The allocated memory is be pinned, so it is safe to use
+-- with FFI calls.
+--
+-- @since 1.2
+genShortByteStringIO :: MonadIO m => Int -> m Word64 -> m ShortByteString
+genShortByteStringIO n0 gen64 = do
+  let !n@(I# n#) = max 0 n0
+      (n64, nrem64) = n `quotRem` 8
```

Comment:
Last time I looked into asm, ``n64 = n `quot` 8; nrem64 = n `rem` 8`` was faster, because GHC was able to replace `quot` and `rem` by bitwise operations, but `quotRem` remained a full-fledged division. Obviously, should not matter, except for particularly short strings.

Reply:
This is correct.

Reply:
Are you serious? How come ghc can't figure that one out? :D 
That is only because `8` is known statically, right? Otherwise it would be pretty sad.

Reply:
GHC is pretty dumb with regards to bit fiddling. But when it is a statically known power of two and GHC is not overwhelmed by returning two values at once, it can cope :P